### DISCUSSION TITLE
OID4VC - VC Data Formats for initial Implementation of OID4VC in keycloak

### DISCUSSION THREAD

**Discussion by francis-pouatcha** [keycloak/keycloak#25536] [Ideas]
This discussion will help narrow down the scope of the work to be done in keycloak with respect to OID4VC.

In the evolving landscape of Self-Sovereign Identity (SSI) and Verifiable Credentials (VC), several competing standards are emerging. These include:

- [Verifiable Credentials based on SD-JWT (SD-JWT-VC)](https://www.ietf.org/archive/id/draft-terbu-oauth-sd-jwt-vc-00.html#name-verifiable-credentials-base): A format with emphasis on selective disclosure and JWT.
- [W3C Verifiable Credentials Data Model (VCDM)](https://www.w3.org/TR/vc-data-model/): The W3C's approach to a standardized VC model. Strongly emphasizing on JSON LD.
- [ISO MDOC - MDLs](https://www.iso.org/standard/69084.html): ISO's Mobile Driver's License standard built on CBOR and including selective disclosure.
- [JWT-VC](https://github.com/decentralized-identity/did-jwt-vc): Another JWT-based VC from the [Decentralized Identity Foundation](https://github.com/decentralized-identity).

Following the recommendations from the [eIDAS 2.0 provisional agreement (pdf)](https://www.europarl.europa.eu/cmsdata/278103/eIDAS-4th-column-extract.pdf) and the corresponding [Architecture and Reference Framework](https://github.com/eu-digital-identity-wallet/eudi-doc-architecture-and-reference-framework), and considering implementation and interoperability efforts performed by [these different organizations and initiatives (Google Sheet - OID4VC Profiles)](https://docs.google.com/spreadsheets/d/1s6REK5eNAb3GSElID0J02_TtbuI2Exd9z-CLdLx0emk/edit?usp=sharing), we advise the keycloak implementation to start with the credential formats:
*  SD-JWT-VC and 
*  ISO mDLs. 

We welcome feedback on this approach, particularly insights into the implementation and integration of these standards. Any additional perspectives would also be greatly appreciated.

These topic is important as it will help us nail down libraries to be used for the handling of those formats in the keycloak source code.

Thank you for contributing to this important discussion!

**Comment by francis-pouatcha** [comment]
We could use OID4VC as top level protocol name. Means e.g.

- Breakdown in VCI, VP and SIOP

oid4vc/oid4vci/[vcdm|jwt-vc] +  ldproof + eddsa
oid4vc/oid4vp/vcdm
oid4vc/siopv2/vcdm

oid4vc/oid4vci/sd-jwt -> jwt+jws + EC256 (ecdsa)
oid4vc/oid4vp/sd-jwt
oid4vc/siopv2/sd-jwt

oid4vc/oid4vci/mdoc -> cbor+cose
oid4vc/oid4vp/mdoc
oid4vc/siopv2/mdoc

**Comment by tnorimat** [comment]
EdDSA is in progress https://github.com/keycloak/keycloak/pull/17215

**Comment by wistefan** [comment]
As a proposal for the structure on issuance have look at:
https://github.com/wistefan/keycloak/pull/4

**Reply by francis-pouatcha** [reply]
Thank you @wistefan . Will have a look today.

**Reply by francis-pouatcha** [reply]
Looks great for me. Comment in my review.
